### ISSUE TITLE
🐛 delta `0.12.0` is slower than `0.11.3`

### ISSUE THREAD

**Issue by infokiller** [dandavison/delta#973]
Hi!
I see something similar to #824 (which is fixed in `0.11.2`). I will try to reproduce it in docker and post the output.

**Comment by infokiller** [comment]
I'm unable to repro using something similar to https://github.com/dandavison/delta/issues/824#issuecomment-988948010

**Comment by dandavison** [comment]
Hi @infokiller, are you able to reproduce / measure what you're observing, even outside docker? (I really appreciated the performance work you did in December by the way.)

**Comment by infokiller** [comment]
Oops, I can actually reproduce it in docker, I just misread the output. The behavior is similar to #824: it's slow with many processes running. Here's the full script I used:

```sh
build_oci_img () {
  # RUN apk add --no-cache bash git
  docker build "$@" - <<'EOF'
  FROM archlinux
  RUN pacman -Sy > /dev/null
  RUN pacman -S --noconfirm --needed coreutils curl gcc bc git > /dev/null
  ENV CARGO_HOME=/usr/local
  RUN curl -fsSL 'https://sh.rustup.rs' | sh -s -- -y --profile minimal
  RUN /usr/local/bin/cargo install --version 0.11.3 git-delta && cp /usr/local/bin/delta /usr/local/bin/delta-0.11.3
  RUN /usr/local/bin/cargo install --version 0.12.0 git-delta && cp /usr/local/bin/delta /usr/local/bin/delta-0.12.0
  WORKDIR /repo
EOF
}

build_oci_img
docker run --rm -i "$(build_oci_img -q)" bash <<'EOF'
benchmark-command() {
  local num_runs="${1:-100}"
  shift
  local before
  before=$(($(date +%s%N) / 1000000))
  printf 'Running command "%s" %d times\n' "$*" "${num_runs}"
  for _ in $(seq 1 "${num_runs}"); do
    eval -- "$@" > /dev/null
  done
  local after
  after=$(($(date +%s%N) / 1000000))
  local per_run
  per_run=$(bc < <(printf 'scale=2; %d/%d\n' $((after - before)) "${num_runs}"))
  echo "${per_run} ms per run ($((after - before)) ms total)"
}

benchmark-delta-versions() {
  for v in 0.11.3 0.12.0; do
    echo "Benchmarking version: $v"
    delta-${v} --version
    benchmark-command 10 "git diff --no-index a b | delta-${v} >| delta-${v}-result"
  done
}

main() {
  echo aaa > a
  echo aab > b
  git diff --no-index a b | delta
  benchmark-delta-versions
  echo "Spinning up 5000 processes"
  for i in {1..5000}; do
    nohup sleep 30 & &> /dev/null
  done
  echo "Process counts:"
  ps --no-headers -e -o fname | sort | uniq -c
  benchmark-delta-versions
}

main
EOF
```

Which results in:
```
Spinning up 5000 processes
Process counts:
      1 bash
      1 ps
   5000 sleep
      1 sort
      1 uniq
Benchmarking version: 0.11.3
delta 0.11.3
Running command "git diff --no-index a b | delta-0.11.3 >| delta-0.11.3-result" 10 times
76.70 ms per run (767 ms total)
Benchmarking version: 0.12.0
delta 0.12.0
Running command "git diff --no-index a b | delta-0.12.0 >| delta-0.12.0-result" 10 times
419.70 ms per run (4197 ms total)
```

**Comment by dandavison** [comment]
@infokiller these docker-based repro scripts are works of art.

@th1000s what do you think is the right course of action here?

In #977 I reverted 15d7bb8cf0b3973719a97fa206ac94e5b46e0ede. Superficially at least, that seems to solve the performance problem as we probably expected (`dev` is #977):

```
Benchmarking version: 0.11.3
delta 0.11.3
Running command "git diff --no-index a b | delta-0.11.3 >| delta-0.11.3-result" 10 times
74.00 ms per run (740 ms total)
Benchmarking version: 0.12.0
delta 0.12.0
Running command "git diff --no-index a b | delta-0.12.0 >| delta-0.12.0-result" 10 times
302.50 ms per run (3025 ms total)
Benchmarking version: dev
delta 0.12.0
Running command "git diff --no-index a b | delta-dev >| delta-dev-result" 10 times
61.40 ms per run (614 ms total)
```

**Comment by infokiller** [comment]
Thanks for the kind words @dandavison 
Some parts are hard to read and it's more common to actually use a separate Dockerfile but this way it's self contained.
Let me know if something isn't clear or you want me to make any changes (for example pass the versions that are benchmarked as an argument to the script).
I can also confirm I get the same results you see when running:

```sh 
curl -fsSL 'https://raw.githubusercontent.com/dandavison/delta/973-processes-performance/etc/examples/973-processes-performance-bug' | sh -s --
```

**Comment by th1000s** [comment]
I think this is in line with the general slowdown when listing a lot of processes (and/or threads, hence the `-T` argument for `ps`).

When I compare the timing of `ps -A -T` vs. `delta` which only queries processes (#979), I observe that both take about 80ms with about 3000 processes/threads. After spawning 5000 more sleep processes both now take about 300ms. 

This is caused by the `git diff` process exiting before delta has a chance to find it, so it will query all processes to build the whole process tree (but still not find its caller). When delta is configured as a pager this query usually does not happen because the parent process is found instantly.

**Comment by infokiller** [comment]
Is it possible to opt out from querying the processes?

**Comment by dandavison** [comment]
> Is it possible to opt out from querying the processes?

It's not currently, but I'd like to carry on improving this situation. For now, @infokiller, I'm assuming you have a workaround for your personal work? E.g. something like [2db5463](https://github.com/dandavison/delta/pull/977/commits/2db54632c73d2e19e4fcb4e0e74f2ad82c390738).

> When delta is configured as a pager this query usually does not happen because the parent process is found instantly.

Out of interest, what commands are you running in practice that result in you seeing this problem? I see your benchmarking command is

```
git diff --no-index a b | delta-${v}
```

is it `delta a b` which has caused you to notice this? We can add a flag to control the full process scan, and we can also look for places where we shouldn't be doing it.

**Comment by infokiller** [comment]
@dandavison I notice it when running `git diff`. My git pager is defined like this:
```gitconfig
[core]
	pager = "delta 2> /dev/null || { { { diff-so-fancy || cat ; } 2> /dev/null ; } | less --tabs=4 -RFX -- ; }"
```

The expression is complex because I want it to have a fall back if delta is not installed.

**Comment by infokiller** [comment]
> > Is it possible to opt out from querying the processes?
> 
> It's not currently, but I'd like to carry on improving this situation. For now, @infokiller, I'm assuming you have a workaround for your personal work? E.g. something like [2db5463](https://github.com/dandavison/delta/pull/977/commits/2db54632c73d2e19e4fcb4e0e74f2ad82c390738).

My current workaround is to install `0.11.3` using `cargo install --version 0.11.3 --locked --all-features`

**Comment by dandavison** [comment]
> My git pager is defined like this:

OK, thanks -- that's interesting. (I'm not opposed to adding more CLI control but) I wonder whether it might be possible to adjust the parent-process-finding code so that it finds the relevant process without the full process scan, when using a shell command like yours as `core.pager`. @th1000s may have some ideas here.

It's not immediately obvious to me why the git process isn't found quickly with delta involved in that shell construction. A very quick experiment doesn't make me understand that there is anything tricky about the process tree:


```
bash -c 'sleep 1 2> /dev/null || { { { sleep 2 || cat ; } 2> /dev/null ; } | sleep 3 ; }' &; pstree > /tmp/procs
```

```
 | | |-+= 63043 ddavison bash -c sleep 1 2> /dev/null || { { { sleep 2 || cat ; } 2> /dev/null ; } | sleep 3 ; }
 | | | \--- 63045 ddavison sleep 1
```

**Comment by th1000s** [comment]
This (interesting!) delta-then-fallback pager setup works fine for me. The logic delta uses to find the parent `git` process is to first check the parent process, this works when configured like `pager = "delta"`. Then it looks at about 30 pids (-10/+20) and builds a process subtree. This finds sibling processes created by `git show | delta` (unless the command on the left side of the pipe exits to quickly), but also helps in cases where there are intermediary processes, e.g. the `/bin/sh -c`. Note that the `delta` pid 4690 is exactly 2 higher than the one of `git`.

````
$ ps auxf # relevant parts only
th1000s    4303    \_ -zsh
th1000s    4688        \_ /usr/bin/git show
th1000s    4689            \_ /bin/sh -c delta || { { { diff-so-fancy || cat ; [ ... ]
th1000s    4690                \_ delta
th1000s    4694                    \_ less --RAW-CONTROL-CHARS --quit-if-one-screen
````

If nothing is found a full process scan is started, I wonder how you run into that. Do you use a kernel which randomizes pids? One extra fallback might involve just looking at a few more direct parents.

> opt out from process querying

While easy to add, making it configurable is a bit more complicated and might have to use an env variable.

**Comment by infokiller** [comment]
> Do you use a kernel which randomizes pids? One extra fallback might involve just looking at a few more direct parents.

I'm not sure, I use [linux-hardened](https://github.com/anthraxx/linux-hardened) but couldn't find documentation about this specific feature. When I run `sleep 3 & pstree -p $$` I get:

```
zsh(4009343)─┬─pstree(4009752)
             └─sleep(4009751)
```

I should note that `pstree -p $$` is also very slow for me. strace shows a lot of time is spent on `newfstatat(AT_FDCWD, "/proc/X/ns/Y"`.

**Comment by infokiller** [comment]
[This issue](https://bugs.launchpad.net/ubuntu/+source/psmisc/+bug/1764458) also reports pstree being very slow with many processes/threads.

**Comment by th1000s** [comment]
I can only speculate, but it seems part of the hardening process makes some syscalls slower. Even with the wrapper script the caller should be found after looking up at most 30 processes.

With #984 delta now also looks at the grandparent, so this should be down to 2 processes.

**Comment by dandavison** [comment]
Fixed by #984 (and released)
